src/service/notification_service/nostr.rs

```diff
@@ -60,12 +97,220 @@ impl NotificationJsonTransportApi for NostrClient {
-            self.client
-                .send_private_msg(public_key, message, None)
-                .await?;
+            if let Some(relay) = &recipient.nostr_relay {
+                self.client
+                    .send_private_msg_to(vec![relay], public_key, message, None)
+                    .await?;
+            } else {
+                self.client
+                    .send_private_msg(public_key, message, None)
+                    .await?;
+            }
         } else {
-            error!("Nostr npub not found");
+            error!(
+                "Try to send Nostr message but Nostr npub not found in contact {}",
+                recipient.name
+            );
         }
         Ok(())
     }
 }
+
+#[derive(Clone)]
+pub struct NostrConsumer {
+    client: NostrClient,
+    event_handlers: Arc<Vec<Box<dyn NotificationHandlerApi>>>,
+    contact_service: Arc<dyn ContactServiceApi>,
+}
+
+impl NostrConsumer {
+    #[allow(dead_code)]
+    pub fn new(
+        client: NostrClient,
+        contact_service: Arc<dyn ContactServiceApi>,
+        event_handlers: Vec<Box<dyn NotificationHandlerApi>>,
+    ) -> Self {
+        Self {
+            client,
+            event_handlers: Arc::new(event_handlers),
+            contact_service,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn start(&self) -> Result<JoinHandle<()>> {
+        // move dependencies into thread scope
+        let client = self.client.clone();
+        let event_handlers = self.event_handlers.clone();
+        let contact_service = self.contact_service.clone();
+
+        // subscribe only to private messages sent to our pubkey
+        client
+            .subscribe(Filter::new().pubkey(client.public_key).kind(Kind::GiftWrap))
+            .await
+            .expect("Failed to subscribe to Nostr events");
+
+        // run subscription in a tokio task
+        let handle = tokio::spawn(async move {
+            client
+                .client
+                .handle_notifications(|note| async {
+                    if let Some((envelope, sender, _event_id)) = client.unwrap_envelope(note).await
+                    {
+                        // TODO: Check if we already processed this event via event_id
+                        // We only want to handle events from known contacts
+                        if let Ok(sender) = sender.to_bech32() {
+                            trace!("Received event: {envelope:?} from {sender:?}");
+                            if contact_service.is_known_npub(sender.as_str()).await? {
+                                trace!("Received event: {envelope:?} from {sender:?}");
+                                handle_event(envelope, &event_handlers).await?;
+                            }
+                        }
+                    };
+                    Ok(false)
+                })
+                .await
+                .expect("Nostr notification handler failed");
```

Comment:
Should this restart, or something along those lines if it fails, or is it fine, if the app crashes in this case?
We should probably think about that for the dht part as well :thinking: